crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+        &*((self.instances_start + (index * self.instance_size)) as *const Instance)
+    }
+
+    unsafe fn get_location(&self, addr: usize) -> Option<AddressLocation> {
+        // Check for a memory location
+        if addr >= self.memories_start && addr < self.memories_end {
+            let index = (addr - self.memories_start) / self.memory_size;
+            let memory_index = index % self.max_memories;
+            let memory_start = self.memories_start + (index * self.memory_size);
+            let page_index = (addr - memory_start) / WASM_PAGE_SIZE;
+            let instance = self.get_instance(index / self.max_memories);
+
+            let init_page_index = instance
+                .memories
+                .get(
+                    DefinedMemoryIndex::from_u32(memory_index as u32)
+                        .try_into()
+                        .unwrap(),
+                )
+                .and_then(|m| {
+                    if page_index < m.size() as usize {
+                        Some(page_index)
+                    } else {
+                        None
+                    }
+                });
+
+            return Some(AddressLocation::MemoryPage {
+                page_addr: (memory_start + page_index * WASM_PAGE_SIZE) as _,
+                len: WASM_PAGE_SIZE,
+                instance,
+                memory_index,
+                page_index: init_page_index,
+            });
+        }
+
+        // Check for a table location
+        if addr >= self.tables_start && addr < self.tables_end {
+            let index = (addr - self.tables_start) / self.table_size;
+            let table_start = self.tables_start + (index * self.table_size);
+            let table_offset = addr - table_start;
+            let page_index = table_offset / self.page_size;
+
+            return Some(AddressLocation::TablePage {
+                page_addr: (table_start + (page_index * self.page_size)) as _,
+                len: self.page_size,
+            });
+        }
+
+        // Check for a stack location
+        if addr >= self.stacks_start && addr < self.stacks_end {
+            let index = (addr - self.stacks_start) / self.stack_size;
+            let stack_start = self.stacks_start + (index * self.stack_size);
+            let stack_offset = addr - stack_start;
+            let page_offset = (stack_offset / self.page_size) * self.page_size;
+
+            return Some(AddressLocation::StackPage {
+                page_addr: (stack_start + page_offset) as _,
+                len: self.page_size,
+                index,
+                guard_page: stack_offset < self.page_size,
+            });
+        }
+
+        None
+    }
+}
+
+unsafe fn wake_guard_page_access(
+    uffd: &Uffd,
+    page_addr: *const u8,
+    len: usize,
+) -> Result<(), String> {
+    // Set the page to NONE to induce a SIGSEV for the access on the next retry
+    region::protect(page_addr, len, region::Protection::NONE)
```

Comment:
Ok maybe I'm just confused how uffd works.

It looks like a uffd backed page is one with read/write permissions, but not physical memory backed? By using NONE here we're asking the kernel to specifically bypass uffd? Then the `reset_guard_page` switches it back to read/write to get faults routed to uffd?

Reply:
I left a previous comment that explains this a little bit and I'm happy to codify this better in the uffd module's comments.

One of the big selling points of uffd is that we don't have to explicitly change the kernel's memory protection levels when instances are allocated/deallocated or when tables/memories grow (or for reused fiber stacks either).

It keeps the regions as entirely read/write and initially the entire memory range is mapped but "missing" (i.e. normally the kernel would service a fault as if the page were all zeroes). When a fault occurs in the region, we can detect if the access was semantically to a "guard" page even though it has read/write access.

The older kernels allowed us to simply wake the faulting thread without servicing the fault and, instead of retrying the fault, it would `SIGBUS` the access (this is behavior Lucet relies on currently to translate the access to a trap).

Newer kernels will retry the fault indefinitely, meaning we *must* service it with either zeroing the page or copying the page from some other VA.  However, we don't want to service it because the access was to a guard page in this case.

So what this does is record the fact a guard page was hit, "temporarily" set the protection to `NONE` to induce a `SIGSEGV` when the kernel retries the fault, and then later on "reset" the protection level back to read/write as some point in the future the page might actually be accessible.  Granted, this implementation could be a little smarter in identifying pages that will *never* become accessible and thus leave them forever with a `NONE` protection level (something I will try to address).

This means we only ever change protection levels on guard page access.  This isn't ideal as malicious modules might intentionally hit guard pages, but there's other mitigations services could do to prevent abuse (i.e. "your module has to sit in the naughty corner because it trapped out of bounds to much").

Reply:
Ok that all definitely makes sense, thanks for explaining!

> One of the big selling points of uffd is that we don't have to explicitly change the kernel's memory protection levels when instances are allocated/deallocated or when tables/memories grow (or for reused fiber stacks either).

Ah right yes, it's all about avoiding those IPIs with cross-cpu TLB flushes?

We do, however, routinely change the page tables but only insofar as adding allocated pages for regions? e.g. filling something in with zeros or from another page.

>  Granted, this implementation could be a little smarter in identifying pages that will never become accessible and thus leave them forever with a NONE protection level (something I will try to address).

Ah yeah this is also something I was wondering about. Presumably the guard page for all async stacks and the 2GB guard region for each memory could always be mapped as `NONE`?

Reply:
The pooling instance allocator now protects statically known guard pages as `NONE` and will only change protections on pages that could potentially be accessible in the future.

The uffd implementation now only receives the faults for the potentially accessible pages as well.